Buzzページのリニューアル

Miya096jp opened the pull request:
<!-- I want to review in Japanese. -->
## Issue

#8895

## 概要
FBCのホームページの紹介・言及記事ページを刷新しました。

このPRでは、大きく分けて次の3つの機能を追加しました。
- 機能1. 年ごとの記事を選択表示
- 機能2. 管理ページでの記事のcrud
- 機能3. urlからの記事のメタデータを自動取得

基本的なcrudに、メタデータ取得機能を追加したかたちです。

### ファイル数について
File Changedが39もありますが、migrationファイル(参照不要)や、ちょっとしたリンク修正などを含むので実質的なレビュー対象のファイル数は15ほどになります。

## Before & After

### Before
マークダウンで直打ちする方法をとっていた

![スクリーンショット 2025-10-18 16 18 20](https://github.com/user-attachments/assets/4809de1c-ed8d-4759-b71d-d70e712639c5)





### After

![スクリーンショット 2025-10-18 16 28 10](https://github.com/user-attachments/assets/0f8d402e-8013-4ecc-9c36-2ffd84069d9d)






## レビューについて

ファイルが多くてわかりにくいので、機能ごとにそれぞれの仕様、見るべきファイル、動作確認手順を整理しました。


## 準備
feature/redesign-buzzes-pageをローカルに取り込む
git fetch origin feature/redesign-buzzes-page
git checkout feature/redesign-buzzes-page

testが通るか確認してください
bin/rails test

---

## 機能1: 年ごとの記事を動的に表示

### 仕様
- 年をクリックして表示する記事を切り替えられる
- デフォルトでは最新年の記事が表示される
- 記事は月毎にグループ化され、新しい月から降順で表示される

### ルーティングとURL構成
```
# config/routes.rb
  get "buzzes/years/:year", to: "buzzes#index", as: :buzzes_year, constraints: { year: /\d{4}/ }
  resources :buzzes, only: %i[index]

# URL例: http://localhost:3000/buzzes/years/2022
```
[補足]view側で`.../2022#buzz`とURLフラグメントが不可されています。

### 動作確認項目
http://localhost:3000/buzzes
- [ ] ページアクセス時にデフォルトで最新年の記事が表示されるか
- [ ] 記事の表示順は適当か(月毎にグループ化され降順で並ぶ)
- [ ] 年をクリックして、表示が切り替わるか

### 主な関連ファイル
app/controllers/buzzes_controller.rb
app/helpers/buzzes_helper.rb
app/views/buzzes/\_buzzes.html.slim
app/views/buzzes/index.html.slim
app/models/buzz.rb

---

## 機能2: 管理ページを刷新しcrudを実装

### 仕様
- url, title, published_atは`presence :true`、memoは任意
- urlのみunique制約あり、title, published_atは重複を許容
- urlのみ編集不可(編集フォームにはurlフィールドを非表示)

### 動作確認項目
メンターページ: http://localhost:3000/mentor/buzzes
- [ ] 記事が降順で表示されているか
- [ ] ページネーションが動作するか
- [ ] title、url、published_at、memoをすべて手入力し、登録できるか
検証用URL: https://bootcamp.fjord.jp/articles/177
- [ ] 編集フォームはurlフィールドが非表示になっているか
- [ ] 既存の記事を編集・削除できるか

### 主な関連ファイル
config/routes/mentor.rb
app/controllers/mentor/buzzes_controller.rb
app/models/buzz.rb
app/views/mentor/buzzes/_form.html.slim
app/views/mentor/buzzes/_mentor_buzzes.html.slim
app/views/mentor/buzzes/edit.html.slim
app/views/mentor/buzzes/index.html.slim
app/views/mentor/buzzes/new.html.slim

---

## 機能3: メタデータの自動取得を実装

### 仕様
基本動作:
1. urlフィールドのみ入力して保存ボタン押下
2. Net::HTTPでサーバーにリクエストしてHTTPを取得
3. Nokogiriでtitleとpublished_atのメタデータを抽出
4. 抽出したデータをDBに保存

- 常に手入力の値が優先して保存される
- メタデータが取得できない場合は手入力を促すメッセージが表示される
[補足]
- titleが取れないケースは既存の1200ほどの記事で1件もなし
- published_atが取得できないケースは10件中1件ほどある

### 動作確認項目
- [ ] URLの値からメタデータを自動取得ができるか
- [ ] `http/https`以外のスキーマのurlは登録できない
検証用URL: `ftp://bootcamp.fjord.jp/articles/178`
- [ ] published_atが取得できない場合の挙動を確認(メッセージ表示 => 手入力)
検証用URL: `https://bootcamp.fjord.jp/articles/180`
[補足]当日の日付を選択するとトップに表示されるので確認しやすいです
- [ ] title、published_atは手入力した値が優先されるか

### 関連ファイル:
機能2と同じ

---

## 本番データのdata migration動作確認

1. Buzzテーブルの中身を削除
```bash
# rails console
Buzz.delete_all
```

2. data migrationファイルを指定して実行
```bash
rake db:migrate:with_data VERSION=20260325014711
```

3. 次のURLにアクセスし本番データ(fjord関連のブログ記事)が表示されていることを確認
http://localhost:3000/buzzes
http://localhost:3000/mentor/buzzes



![スクリーンショット 2026-03-25 14 30 02](https://github.com/user-attachments/assets/a2246ec9-3200-4ce0-95a1-3dde19c87eab)
















<!-- I want to review in Japanese. -->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## リリースノート

* **新機能**
  * 記事一覧ページに年別フィルタリング機能を追加
  * 記事が月別にグループ化されて表示されるようになりました
  * メンター向けの記事管理画面を追加（記事の作成・編集・削除が可能）
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Miya096jp committed:
最小限のseedデータを登録

全件だと一部整形作業が必要なため

Miya096jp committed:
buzzesテーブルにカラムを追加

title, url, memo, published_atカラムを追加
fixturesのpublished_on -> published_atに変更

Miya096jp committed:
routesの設定

Miya096jp committed:
buzzes_controllerとmento/buzzes_controllerを生成

Miya096jp committed:
model/buzz.rbにクラスメソッドを追加

Miya096jp committed:
buzzes_controllerの編集

indexアクションの中身を追加。LPの紹介・言及記事ページに渡すインスタンス変数をセット。

Miya096jp committed:
buzzes#indexアクションへのカスタムルートを設定

get "buzzes/years/:year", to: "buzzes#index"...

Miya096jp committed:
LPの紹介・言及記事ページのviewを編集し、データの表示に成功

Miya096jp committed:
buzzesテーブルからbodyカラムを削除

Miya096jp committed:
mentor/buzzes#indexにkaminariのpagenationを追加/不要なテストファイルを削除

Miya096jp commented on app/models/buzz.rb:
```diff
@@ -1,5 +1,47 @@
-  validates :body, presence: true
+  # validates :body, presence: true
+  validates :published_at, presence: true
+  require 'open-uri'
```
@coderabbitai

引数urlにはフォームから入力される文字列が渡される。(フォームはこれから実装予定)　この前提でもレビュー結果は同じか。

Miya096jp replied:
@coderabbitai

修正済み

Miya096jp committed:
Buzz.doc_from_urlの修正

coderadditaiの指摘

Miya096jp committed:
Mentor::BuzzesController#indexを修正

coderabbitaiの指摘でkaminariのpagenation前にorderメソッドを挿入

Miya096jp committed:
coderabbitai指摘のタイポミスを2箇所修正

Miya096jp committed:
Buzzesテーブルからdefaultオプションを削除

Miya096jp committed:
メンター用のbuzzes管理ページを作成し記事を一覧表示

Miya096jp committed:
管理者用のBuzz管理ページにnewページを作成

Miya096jp committed:
Mentor::BuzzesController#createを作成

Miya096jp committed:
Buzz.doc_from_urlを修正

Miya096jp committed:
管理者用のBuzz管理ページにeditページを作成

Mentor::BuzzesController#edit/update/delete作成
metnor/buzzes/_form.html.slimのurlフィールドをeditページに非表示
mentor/buzzes/edid.html.slimに削除ボタンを追加

Miya096jp committed:
editページ、管理者ページタブのリンク先を修正

いずれもmentor_buzzes_pathへ修正

Miya096jp committed:
Buzzモデルにvalidationを追加

title, url, published_atカラムにそれぞれ追加

Miya096jp commented on db/fixtures/buzzes.yml:
```diff
@@ -1,2 +1,49 @@
+  title: "関西Ruby会議08に参加しましたー！ - utsubo1024’s diary"
+  url: "https://utsubo1024.hatenablog.com/entry/2025/07/02/161748"
+  published_on: 2025-07-02
+
```
@coderabbitai

修正済み

Miya096jp commented on db/migrate/20250828090839_add_columns_to_buzzes.rb:
```diff
@@ -0,0 +1,19 @@
+  def up
+    add_column :buzzes, :url, :string, default: ""
+    add_column :buzzes, :title, :string, default: ""
+    add_column :buzzes, :published_at, :date, default: -> { 'CURRENT_TIMESTAMP' }
```
@coderabbitai

修正済み

Miya096jp commented on app/controllers/buzzes_controller.rb:
```diff
@@ -0,0 +1,10 @@
+
+class BuzzesController < ApplicationController
+  def index
+    @years = Buzz.years
+    year = params[:year] || Buzz.latest_year
+    @year = year
+    @buzzes = Buzz.for_year(year)
+  end
```
@coderabbitai

修正済み

Miya096jp commented on db/schema.rb:
```diff
@@ -112,9 +112,12 @@
+    t.string "title", default: "", null: false
+    t.date "published_at", default: -> { "CURRENT_TIMESTAMP" }, null: false
+    t.text "memo", default: ""
   end
```
@coderabbitai

修正済み

Miya096jp committed:
Mentor::BuzzesControllerに管理者限定アクセスのvalidationを追加

Miya096jp committed:
Buzz.doc_from_urlに例外処理を追加

Miya096jp committed:
update失敗時に422を返却/成功時のメッセージ追加

Miya096jp committed:
Buzzesテーブルのtitle、urlカラムにunique indexを追加

Miya096jp committed:
Buzzモデルの不要なrequire 'open-uri'を削除しrequire 'net/http'を追加

Miya096jp committed:
キャンセル時のリンク先を:backからmentor_buzzes_pathに修正

Miya096jp committed:
CSS idの値、テーブルのカラム名の修正

Miya096jp committed:
published_atの値取得ロジックを変更

他: 値取得失敗時に422返却・errorオブジェクトにメッセージを追加

Miya096jp committed:
不要なviews/buzz/を削除

Miya096jp committed:
titleカラム、urlカラムからunique indexを削除、seedにデータ追加

titleとurlは内容は重複するケースもあるとわかったため

Miya096jp commented on app/views/mentor/buzzes/_mentor_buzzes.html.slim:
```diff
@@ -0,0 +1,27 @@
+    thead.admin-table__header
+      tr.admin-table__labels
+        th.admin-table__label
+          | プラクティス
```
@coderabbitai

修正した

Miya096jp committed:
title, urlからuniqueness: trueを外した

Miya096jp committed:
LPのviewのレイアウトを大まかに整えた

Miya096jp committed:
LPヘッダーの紹介・言及記事のリンクを修正

Miya096jp committed:
LPのviewのレイアウトの大枠を整えた

年・月ごとに記事が降順の表示されるように変更

Miya096jp committed:
helpers/buzzes_helper.rbを作成しヘルパーメソッドを定義

Miya096jp committed:
tailwindCSSで仮のスタイルをあてた

Miya096jp commented on db/migrate/20250830104721_change_default_on_buzzes.rb:
```diff
@@ -0,0 +1,8 @@
+class ChangeDefaultOnBuzzes < ActiveRecord::Migration[6.1]
+  def change
+    change_column_default :buzzes, :url, from: "", to: nil
+    change_column_default :buzzes, :title, from: "", to: nil
+    change_column_default :buzzes, :published_at, from: -> { 'CURRENT_TIMESTAMP' }, to: nil
+    change_column_default :buzzes, :memo, from: "", to: nil
```
@coderabbitat

いずれも追加したので確認を。

Miya096jp commented on app/controllers/mentor/buzzes_controller.rb:
```diff
@@ -0,0 +1,64 @@
+  def create
+    @buzz = Buzz.new(buzz_params)
+
+    doc = Buzz.doc_from_url(buzz_params[:url])
+    date = Buzz.date_from_doc(doc)
+
```
@coderabbitat

buzz.doc_from_urlに例外処理を追加した。

Miya096jp commented on app/models/buzz.rb:
```diff
@@ -1,5 +1,59 @@
+    with: URI::DEFAULT_PARSER.make_regexp(%w[http https]),
+    message:
+    'URLに誤りがあります'
+  }
```
@coderabbitai

一度ユニークインデックスを貼ったが、　その後、urlもtitleも重複するケースがあることがわかったため元に戻した。

Miya096jp commented on db/migrate/20250831234516_remove_unique_index_from_buzzes_title.rb:
```diff
@@ -0,0 +1,5 @@
+class RemoveUniqueIndexFromBuzzesTitle < ActiveRecord::Migration[6.1]
+  def change
+    remove_index :buzzes, :title
+  end
+end
```
@coderabbitai

過去のmigrationファイルを参照しても、disable_ddl_transaction!を使用した例はなかった。また、Buzzモデルのvaridationからはユニーク制約を外した。再チェックを。

Miya096jp committed:
BuzzHelper#buzzes_yearにセーフナビゲーション演算子を追加

Miya096jp committed:
hタグの見出しレベル修正

Miya096jp committed:
BuzzesHelper#buzzes_yearを削除しyearをコントローラーから直接渡すように修正

Miya096jp committed:
Buzz.doc_from_urlにtimeout処理等を追加

Miya096jp committed:
不要なルーティングを削除

Miya096jp committed:
リンク修正

Miya096jp committed:
エラーのrescue後にreturnするように修正

Miya096jp committed:
buzzモデルのインターフェースに対してユニットテストを作成/テストfixturesの作成

Miya096jp committed:
LPのシステムテストを作成

Miya096jp committed:
urlフィールド未入力に対応するガード節を追加

Miya096jp committed:
Buzz管理ページのシステムテストを追加

Miya096jp committed:
fixtures/buzzes.ymlを削除

Miya096jp committed:
Net::HTTPのtimeout設定の秒数を変更

Miya096jp committed:
記事登録・更新時のフラッシュメッセージを修正

Miya096jp committed:
newページへのリンク名を「記事検索」に変更

Miya096jp committed:
Buzzを作成するcreate_buzz_forメソッドをヘルパーに定義し、システムテストで共通化

Miya096jp committed:
Buzzモデルのユニットテストをリファクタリング

Miya096jp committed:
メソッド名をcreate_buzzに修正

Miya096jp committed:
system/buzzes_testをリファクタリング

Miya096jp committed:
system/mentor/buzzes_test.rbのリファクタリング

Miya096jp committed:
createアクションの many lines警告のため、url入力判定をvaridate_url_presenceに切り出した

Miya096jp committed:
format呼び出し時の()内の不要なスペースを削除

Miya096jp committed:
不要なdata migrationファイルを削除

Miya096jp committed:
不要なライブラリ呼び出しを削除

Miya096jp committed:
不要なライブラリの読み込みを削除し、test_helperを読み込み

Miya096jp committed:
slimのコード埋め込み(=)記号抜けを修正

Miya096jp commented:
@machida

お疲れ様です！
お手隙のときにデザインをお願いいたします🙇🏻‍♂️

djskusa-cmd commented:
> **Actionable comments posted: 0**
> 
> <details>
> <summary>🧹 Nitpick comments (8)</summary><blockquote>
> 
> <details>
> <summary>test/system/mentor/buzzes_test.rb (8)</summary><blockquote>
> 
> `12-19`: **セレクタの堅牢性を上げる（data-testid等の導入）**
> 
> `.admin-table__item td:first-child a` はビュー変更に脆弱です。リンクに `data-testid="buzz-title"` を付与し、`all('[data-testid="buzz-title"]').map(&:text)` のように取得すると壊れにくくなります。
> 
> ---
> 
> `27-27`: **WebMockスタブに Content-Type を付与**
> 
> HTMLパース系の実装によっては `Content-Type` なしで失敗・分岐する可能性があります。`text/html; charset=UTF-8` を明示しましょう。
> 
> 
> ```diff
> -  .to_return(status: 200, body: dummy_response)
> +  .to_return(status: 200, headers: { 'Content-Type' => 'text/html; charset=UTF-8' }, body: dummy_response)
> ```
> 
> 
> Also applies to: 61-61
> 
> ---
> 
> `37-45`: **エラーメッセージのI18n依存を弱める**
> 
> `Url`/`URL` 表記や文言がロケールで変わると落ちます。属性名は `Buzz.human_attribute_name(:url)` を用いて期待値を組み立てると堅牢です。
> 
> 
> ```diff
> -  assert_text 'Urlを入力してください'
> +  assert_text "#{Buzz.human_attribute_name(:url)}を入力してください"
> ```
> 
> ---
> 
> `47-53`: **テスト名の否定/肯定が逆**
> 
> 挙動は「編集フォームにURL入力欄は表示しない」なので、テスト名を否定形に修正してください。内容はLGTMです。
> 
> 
> ```diff
> -test 'url field is now shown in buzz edit form' do
> +test 'url field is not shown in buzz edit form' do
> ```
> 
> ---
> 
> `41-41`: **date型入力には明示的に YYYY-MM-DD 文字列を投入**
> 
> `Date.new(...).to_s` でも多くは動きますが、ドライバや型推論差で不安定になることがあります。文字列を直接入れると安定します。
> 
> 
> ```diff
> -      fill_in 'buzz[published_at]', with: Date.new(2025, 9, 10)
> +      fill_in 'buzz[published_at]', with: '2025-09-10'
> ```
> 
> ```diff
> -      fill_in 'buzz[published_at]', with: Date.new(2025, 10, 1)
> +      fill_in 'buzz[published_at]', with: '2025-10-01'
> ```
> 
> ```diff
> -      fill_in 'buzz[published_at]', with: Date.new(2025, 9, 11)
> +      fill_in 'buzz[published_at]', with: '2025-09-11'
> ```
> 
> 
> Also applies to: 66-66, 81-81
> 
> ---
> 
> `55-62`: **スタブの重複をヘルパに抽出**
> 
> 同一HTMLダミーとスタブ設定が複数テストで重複しています。`test/supports/` 配下にヘルパ（例: `stub_fetch_html(url:, title:, published_at_iso:)`）を置くとDRYになります。
> 
> ---
> 
> `92-99`: **削除後の遷移先も検証**
> 
> UIの文言変化だけでなく、削除後に一覧へ戻ることも検証しておくと意図が明確です。
> 
> 
> ```diff
>    accept_confirm do
>      click_link '削除する'
>    end
>    assert_no_text '新しいBuzz'
> +  assert_current_path mentor_buzzes_path
> ```
> 
> ---
> 
> `3-4`: **`test_helper` の二重読み込みを削除**
> 
> test/application_system_test_case.rb の先頭で既に require 'test_helper' があるため、test/system/mentor/buzzes_test.rb の重複行を削除してください。
> 
> ```diff
>  require 'application_system_test_case'
> -require 'test_helper'
> ```
> 
> </blockquote></details>
> 
> </blockquote></details>
> 
> <details>
> <summary>📜 Review details</summary>
> 
> **Configuration used**: Path: .coderabbit.yaml
> 
> **Review profile**: CHILL
> 
> **Plan**: Pro
> 
> <details>
> <summary>📥 Commits</summary>
> 
> Reviewing files that changed from the base of the PR and between 66a8d8301f02ef7018fdce4978e0d487ea7ebcca and ad259f16430d5a13815132b6678ad1733056cdf4.
> 
> </details>
> 
> <details>
> <summary>📒 Files selected for processing (3)</summary>
> 
> * `db/data/20230726024528_create_buzz.rb` (0 hunks)
> * `test/models/buzz_test.rb` (1 hunks)
> * `test/system/mentor/buzzes_test.rb` (1 hunks)
> 
> </details>
> 
> <details>
> <summary>💤 Files with no reviewable changes (1)</summary>
> 
> * db/data/20230726024528_create_buzz.rb
> 
> </details>
> 
> <details>
> <summary>🚧 Files skipped from review as they are similar to previous changes (1)</summary>
> 
> * test/models/buzz_test.rb
> 
> </details>
> 
> <details>
> <summary>🧰 Additional context used</summary>
> 
> <details>
> <summary>📓 Path-based instructions (2)</summary>
> 
> <details>
> <summary>**/*.rb</summary>
> 
> 
> **⚙️ CodeRabbit configuration file**
> 
> > `**/*.rb`: # refactoring
> > - まずFat Controllerを避け、次にFat Modelを避ける。
> > - Serviceクラスの乱用を避ける。
> > - controller concernを作ろうとしたらPORO(Plain Old Ruby Object)やActiveRecordモデルでの実装で代替できないか検討する。
> > # Rails Patterns
> > - ViewHelperにメソッドを実装する時にはまずDecoratorパターンを使うことを検討する。（active_decorator gemを導入しているのでそれを使う）
> > - 複雑なActiveRecordクエリがあり、再利用できそうな場合はQueryObjectパターンを検討する。（rails-patterns gemを導入しているのでそれのQuery機能を使う）
> > - Viewにpartialを作る場合はViewComponentを使うことを検討する。
> > - 複数のActiveRecordモデルを操作する１つの責務がある時や外部APIとやりとりする処理がある場合にはInteractorオブジェクトパターンを検討する。（interactor gemを導入しているのでそれを使う）
> > - 複数のInteractorを実行するような処理がある場合Organizerオブジェクトパターンを検討する。（interactor gemを導入しており、その中にOrganizerの機能があるので使う）
> 
> Files:
> - `test/system/mentor/buzzes_test.rb`
> 
> </details>
> <details>
> <summary>test/**/*</summary>
> 
> 
> **⚙️ CodeRabbit configuration file**
> 
> > `test/**/*`: # Test
> > - TestCase名は英語で書く。
> > - どうしても避けられない時以外にsystem testでsleepは使わない。
> > # Unit Test
> > model, helper, decorator, view_componentについてはメソッドを追加した場合は必ず対応したUnit TestのTestCaseを1つは書く。
> 
> Files:
> - `test/system/mentor/buzzes_test.rb`
> 
> </details>
> 
> </details><details>
> <summary>🧠 Learnings (4)</summary>
> 
> <details>
> <summary>📓 Common learnings</summary>
> 
> ```
> Learnt from: Miya096jp
> PR: fjordllc/bootcamp#9102
> File: db/fixtures/buzzes.yml:0-0
> Timestamp: 2025-09-04T01:27:23.422Z
> Learning: The buzz_controller.rb and related data migration files were deleted as part of the Buzz page redesign, not just modified. The old single buzz page was completely replaced with a new buzzes (plural) listing page with year-based navigation and metadata structure.
> ```
> 
> ```
> Learnt from: Miya096jp
> PR: fjordllc/bootcamp#9102
> File: db/fixtures/buzzes.yml:0-0
> Timestamp: 2025-09-04T01:27:23.422Z
> Learning: The buzz_controller.rb was indeed deleted, but there are still two remaining items from the old buzz single page implementation that need cleanup: the data migration file db/data/20230726024528_create_buzz.rb and the old buzz route in config/routes.rb line 136. These should be updated/removed to complete the buzz page redesign migration.
> ```
> 
> ```
> Learnt from: Miya096jp
> PR: fjordllc/bootcamp#9102
> File: db/migrate/20250831044340_add_unique_index_to_buzzes_url_and_title.rb:3-4
> Timestamp: 2025-09-12T00:33:15.006Z
> Learning: PR #9102では、Buzzテーブルにユニークインデックスを追加後、同じPR内で削除するというマイグレーションの流れが採用された。これは業務要件でURL/タイトルの重複が許容されるためユニーク制約が不適切だったことによる修正。
> ```
> 
> ```
> Learnt from: Miya096jp
> PR: fjordllc/bootcamp#9102
> File: app/views/mentor/buzzes/edit.html.slim:1-0
> Timestamp: 2025-09-12T01:00:58.415Z
> Learning: app/views/mentor/buzzes/edit.html.slim では `- title` と `- set_meta_tags` の両方が正しく設定されており、タイトルは削除されていない。変更内容はメタディスクリプションの追加のみ。
> ```
> 
> ```
> Learnt from: Miya096jp
> PR: fjordllc/bootcamp#9102
> File: db/migrate/20250831235134_remove_unique_index_from_buzzes_url.rb:1-5
> Timestamp: 2025-09-04T01:56:02.226Z
> Learning: ExternalEntryモデルはbuzzesテーブルとは無関係で、ユーザーのRSSフィードから外部記事を取得・保存する機能を持つ。ExternalEntry.find_by(url: ...)はbuzzesのURL重複制約とは関係ない。
> ```
> 
> ```
> Learnt from: Miya096jp
> PR: fjordllc/bootcamp#9102
> File: app/models/buzz.rb:0-0
> Timestamp: 2025-09-04T01:41:06.709Z
> Learning: Buzzモデルでは、URLとタイトルの重複が業務上発生する可能性があるため、uniqueness制約は使用しない。一度ユニークインデックスを追加したが、重複するケースがあることが判明したため削除された。
> ```
> 
> </details>
> <details>
> <summary>📚 Learning: 2025-09-04T01:27:23.422Z</summary>
> 
> ```
> Learnt from: Miya096jp
> PR: fjordllc/bootcamp#9102
> File: db/fixtures/buzzes.yml:0-0
> Timestamp: 2025-09-04T01:27:23.422Z
> Learning: The buzz_controller.rb and related data migration files were deleted as part of the Buzz page redesign, not just modified. The old single buzz page was completely replaced with a new buzzes (plural) listing page with year-based navigation and metadata structure.
> ```
> 
> **Applied to files:**
> - `test/system/mentor/buzzes_test.rb`
> 
> </details>
> <details>
> <summary>📚 Learning: 2025-09-04T01:27:23.422Z</summary>
> 
> ```
> Learnt from: Miya096jp
> PR: fjordllc/bootcamp#9102
> File: db/fixtures/buzzes.yml:0-0
> Timestamp: 2025-09-04T01:27:23.422Z
> Learning: The buzz_controller.rb was indeed deleted, but there are still two remaining items from the old buzz single page implementation that need cleanup: the data migration file db/data/20230726024528_create_buzz.rb and the old buzz route in config/routes.rb line 136. These should be updated/removed to complete the buzz page redesign migration.
> ```
> 
> **Applied to files:**
> - `test/system/mentor/buzzes_test.rb`
> 
> </details>
> <details>
> <summary>📚 Learning: 2025-09-12T01:00:58.415Z</summary>
> 
> ```
> Learnt from: Miya096jp
> PR: fjordllc/bootcamp#9102
> File: app/views/mentor/buzzes/edit.html.slim:1-0
> Timestamp: 2025-09-12T01:00:58.415Z
> Learning: app/views/mentor/buzzes/edit.html.slim では `- title` と `- set_meta_tags` の両方が正しく設定されており、タイトルは削除されていない。変更内容はメタディスクリプションの追加のみ。
> ```
> 
> **Applied to files:**
> - `test/system/mentor/buzzes_test.rb`
> 
> </details>
> 
> </details>
> 
> </details>
> 
> <details>
> <summary>⏰ Context from checks skipped due to timeout of 90000ms. You can increase the timeout in your CodeRabbit configuration to a maximum of 15 minutes (900000ms). (1)</summary>
> 
> * GitHub Check: build_and_test
> 
> </details>
> 
> </details>
> 
> <!-- This is an auto-generated comment by CodeRabbit for review status -->

machida commented:
@Miya096jp デザイン着手します！！

machida committed:
buzzes周りのデザインを整えた

machida committed:
buzz管理画面にある不要なIDとクラスを削除

machida committed:
メンターbuzzes管理画面の改善

machida committed:
モバイル用のBuzz年別フィルターの選択ロジックを修正

machida commented:
@Miya096jp デザイン入れました！！テストが落ちたら修正しますー

machida committed:
Buzz登録時のURL取得失敗をハンドリング

machida committed:
buzzesテストのフィクスチャ日付の検証を修正

machida committed:
reverse.eachをreverse_eachに変更

machida commented on test/system/buzzes_test.rb:
```diff
@@ -0,0 +1,57 @@
+      assert_text '2024-12-31の記事'
+      assert_text '2024-12-01の記事'
+      assert_no_text '2024-11-20の記事'
+    end
```
@coderabbitai 修正完了

machida committed:
RuboCopが指摘した規約違反の箇所を修正

machida committed:
Buzzesページリデザインに伴うテストを修正

machida committed:
タイポ修正

Miya096jp committed:
Google拡張クライアント用のAPIエンドポイントを作成

rack-cors gemでCORS設定を追加

Miya096jp committed:
レビュー依頼前の最終チェック

定数の利用やメソッド名の微修正など、細かな変更を行なった

smallmonkeykey commented:
@Miya096jp 
すみません、現在他のレビューを担当しているため他の方にお願いしていただいても良いですか？
確認したところFjord Choiceでは、`レビュー依頼はありません`で外れていて申し訳ないです🙇‍♂️

Miya096jp commented:
@matuaya 

お疲れ様です！
急ぎませんので、こちらのレビューをお願いできませんでしょうか？
ご検討よろしくお願いいたします🙇🏻‍♂️

matuaya commented:
@Miya096jp 
お疲れ様です！
レビュー依頼していただきありがとうございます☺️
もしかしたら１０日以上かかってしまうかもしれないのですが大丈夫でしょうか？🙇‍♀️💦

Miya096jp committed:
coderabbitの指摘に対応

クローズした#9283の指摘を受けてエラーハンドリングを追加
StrongParameterにurlを含めるように変更

matuaya commented:
@Miya096jp 
ありがとうございます🙇‍♀️
レビュー承りました☺️

Miya096jp committed:
destroyアクションを追加

Miya096jp committed:
checkアクションのレスポンスに@buzzを含めるように修正

Miya096jp committed:
api/buzzes/lookups_controllerにshowアクションを定義

RESTfulな構成にするためにコントローラーを分割しcheckアクションを移設・修正

matuaya commented on app/controllers/mentor/buzzes_controller.rb:
```diff
@@ -0,0 +1,89 @@
+  def create
+    @buzz = Buzz.new(buzz_params)
+
+    return render :new, status: :unprocessable_entity unless validate_url_presence
```
無効なURLをフォームで登録しようとすると例外が発生してページがエラーになってしまうので、URLの有無確認と一緒にURLが有効であることもここでバリデーション検証した方がよいかもしれないので検討いただけると嬉しいです☺️🙏

その場合Controllerでのバリデーションのハードコーディング記述が増えてしまいそうなので、Modelのバリデーションを使えないかなと探してみたところ、こんなドキュメントを見つけました！
https://railsguides.jp/active_record_validations.html#on

ModelのURLのバリデーション部分でこんな風に書くと
```
  validates :url, presence: true, format: {
                                    with: URI::DEFAULT_PARSER.make_regexp(%w[http https]),
                                    message: 'URLに誤りがあります'
                                  },
                  uniqueness: { message: 'このURLはすでに登録されています' }, on: :some_custom_name
```
Contollerでこのように任意のタイミングでバリデーションを行えるようです。
```
return render :new, status: :unprocessable_entity unless @buzz.valid?(:some_custom_name)
```
これだとURLの有無と有効性が同時に検証でき、Contollerでバリデーション処理をハードコーディングする必要がなくなるかなと思います。
ただ実際のところ、Controllerでこの方法を使うのか適切かどうか私には判断できないので、あくまで参考までにとどめていただいて大丈夫です🙇‍♀️💦

matuaya replied:
なるほどです！
参考サイトも共有していただきありがとうございます☺️🙏
コントローラーにまとめることで流れが明確になるのは確かに利点が大きいと思いました！
勉強になりました☺️

matuaya commented on app/models/buzz.rb:
```diff
@@ -1,5 +1,74 @@
+    def date_from_doc(doc)
+      date = doc.at_css('meta[property="article:published_time"]')&.[]('content')&.slice(0, 10)
+      date ? Date.parse(date) : nil
+    end
```
正直どこまでの処理をModelに含めて良いのか私には分からない部分があるのですが、なんとなくここのHTTP処理がBuzzのModelの責務から少し離れているような気がするのですがどうでしょうか🙏💦
サービスクラスという手もありそうですが、処理も多くないのでこのままでもいいかもしれないとも思います💦

matuaya replied:
こちらも参考サイトの共有ありがとうございます🙏

モデルに置く判断の背景を丁寧に説明していただき、とても勉強になりました！
私もその方針で良いと思いました☺️🙏

matuaya commented on db/migrate/20250831044340_add_unique_index_to_buzzes_url_and_title.rb:
以下の4つのmigrationファイル（
`20250831044340_add_unique_index_to_buzzes_url.rb`
`20250831234516_remove_unique_index_from_buzzes_title.rb`
`20250831235134_remove_unique_index_from_buzzes_url.rb`
`20251022075147_add_unique_index_to_buzzes_url.rb 07-12-12-337.rb`）

は最終的に`add_index :buzzes, :url, unique: true`のみを適用していると思います。
このファイル１つにまとめてしまってもいいんじゃないかなと思うので検討していただけると嬉しいです☺️🙏

matuaya reviewed (commented):
大変遅くなってしまい申し訳ありません🙇‍♀️
いくつかコメントさせていただいたので検討していただけると嬉しいです🙏

Miya096jp committed:
buzzesコントローラーのルーティングを単一リソースに変更

- destroyアクションのリソースをurlでクエリーするため
-
destroyアクションのelse句の構文エラーを修正し、正しく404が返却されるようにした

Miya096jp committed:
require_admin_or_mentor_login_for_apiメソッドをconcernsに移動

Miya096jp committed:
urlの未入力と無効なスキーマに対するエラーハンドリングを追加

Miya096jp committed:
不要なマイグレーションファイルを削除

Miya096jp committed:
API::Buzzes::LookupsControllerを~::BuzzControllerに変更

- lookups_controllerからbuzz_controllerに名前変更/クライアントも合わせて修正し動作確認済み

matuaya reviewed (dismissed):
Approveさせていただきます☺️

丁寧に説明いただきありがとうございます！大変勉強になりました😊🙏

Miya096jp committed:
buzzesのクライアントページ(日報・言及記事)をwelcomeページの配下に移動

db/fixtures/buzzes.ymlのデータ件数を削減

Miya096jp committed:
例外処理追加

Miya096jp committed:
パスヘルパーの修正

Miya096jp committed:
Buzzのtestにsetupを追加

CIのテストのデータベースにdb/fixturesのデータが混入してテストが落ちるため、テスト実行直前にデータベースをdelete_allするsetupを追加した。

Miya096jp committed:
urlカラムににunique制約を追加

合わせてseedsのurl重複を修正

Miya096jp committed:
BuzzHelperを変更し、呼び出し側も修正

直前のコミット(c92c227)に伴う変更

Miya096jp committed:
.gitignoreに.DS_Storeを追加

Miya096jp commented on app/models/buzz.rb:
```diff
@@ -1,5 +1,82 @@
+  validates :published_at, presence: true
+  validates :url, presence: true, uniqueness: { message: 'はすでに登録されています' }
+  validate :url_format
+
```
@coderabbitai
コメントアウト解除済みです。

Miya096jp commented:
@okuramasafumi 

お疲れ様です！
レビューをお願いいたします。

Miya096jp committed:
コントローラーを統合

Miya096jp committed:
lint, Gemfile更新

Miya096jp committed:
buzzes_controllerのvalidate_url等のメソッドをbuzz_modelに移動

その他:
- flashメッセージの文言を修正
- for_yearメソッドをリファクタリング

Miya096jp committed:
destroyメソッドが何らかの理由でfalseになった場合にエラーが握りつぶされる問題に対処

Miya096jp committed:
不要な変数への再代入を削除

Miya096jp committed:
google拡張のcors設定許可についてコメントを追記

Miya096jp committed:
ルーティングの記述をよりシンプルに修正

Miya096jp committed:
テストケース毎にヘルパーメソッドを呼び出してテスト用のインスタンスを作成していたが、すべてのテストケースでfixtureを使うように修正した

Miya096jp commented on app/controllers/api/buzzes_controller.rb:
```diff
@@ -0,0 +1,55 @@
+  def destroy
+    @buzz = set_buzz
+    if @buzz
+      @buzz.destroy
```
ご指摘ありがとうございます。削除失敗時にfalseが返ることを考慮できていませんでした。`@buzz && @buzz.destroy`に修正しました。

Miya096jp commented on app/controllers/mentor/buzzes_controller.rb:
```diff
@@ -0,0 +1,93 @@
+    true
+  end
+
+  def validate_url
```
こちらもモデルに移しました。

Miya096jp commented on app/controllers/mentor/buzzes_controller.rb:
```diff
@@ -0,0 +1,93 @@
+      if date.present?
+        @buzz.published_at = date
+      else
+        @buzz.errors.add(:published_at, 'を入力してください')
```
確かに、「取得できない」ことを伝えるのが正しいですね。
その上で自然にユーザーの入力を促すような書き方に変えてみました。

「published_at(記事の投稿日)の値が取得できませんでした。任意の日付を入力してください」

Miya096jp commented on test/models/buzz_test.rb:
```diff
@@ -0,0 +1,95 @@
+
+  setup do
+    Buzz.delete_all
+  end
```
fixtureを使う代わりにテストケース毎に明示的に作成した方が読みやすいかなと思ったのですが、その場合、CIでfixture読み込み時のDB初期化が実行されないことが原因で、seedsのデータがCIのテスト用DBに混入する状態になっていました。それに対処するためにdelete_allを使っていました。

fixtureを使ったスタンダードな書き方に変更し、こちらは削除しました。併せてテストヘルパーも削除しています。特に読みにくくなることもなく、元の書き方は考えすぎだったかなと思います。

okuramasafumi reviewed (changes requested):
概ね大丈夫そうでした、指摘した箇所だけ修正と確認をお願いします！

Miya096jp committed:
Buzz#yearsメソッドから不要なセーフナビゲーション演算子を削除

Miya096jp committed:
populate_buzz_from_docメソッドとvalidate_urlメソッドをインスタンスメソッドに修正

Miya096jp commented on app/models/buzz.rb:
```diff
@@ -1,5 +1,114 @@
+        return false
+      end
+      true
+    end
```
ありがとうございます！基本を再確認できました。selfについては、単にプロパティを参照する場合は省略できるけど、値を代入するときは必須ということですね。

Miya096jp commented:
@okuramasafumi 

メンションが抜けており、申し訳ありません。お気づきいただいてありがとうございます、よく理解できました🙏　孫ブランチを切って作業に取り掛かりたいと思います。

こちらから派生した作業なので、引き続きレビューをお願いしてよろしいでしょうか？　終わり次第メンションさせていただきますので、ご確認をよろしくお願いいたします🙇🏻‍♂️

Miya096jp committed:
本番データのcsvファイルとdata migrationファイルを作成

csvファイルは別PRで削除予定

okuramasafumi reviewed (approved):
大丈夫そうです、データ追加のマイグレーションは約1000件の`INSERT`処理が走るので重いですが、そういうものだと割り切れば良いと思います（ただし実行するタイミングによってはアプリが重くなりそう）。

okuramasafumi commented:
@Miya096jp 遅くなってすみません、レビュー完了しました。あとはコンフリクトを解消する必要がありますね。

---

今回のプルリクエストはコミットの数が100件を超えており、追加行数も2000行を超えてしまっています（ただし、追加行数の中身としてはCSVが多いので、その分は割り引いて考える必要はあります）。
実際の業務では、コミットが100を超えることはかなり稀です。理由としては、コミットを作業単位として見なす場合（そして見なすべきです）、100の作業がまとまっている（何もかもがそのまとまりに対して行われてしまう）状況は明らかに望ましくないためです。実務的には、コミットを統合して数を減らし、その上で論理的に意味のある単位に再分割することでレビューが容易になり、巻き戻し（revert）も容易になります。
そのためには、`git rebase`や`git commit --amend`のようなコマンドを駆使することになります。とはいえ、コミットが100ある状況から整理するのはベテランでもかなり大変です。そのため、今回はコミットが100あるとしてもそのまま進行するほうがよい、という判断をしています。

実務でこのようなことが起こるときには大抵同じような背景があります。それは、

1. 要求や要件が曖昧または抽象的すぎる
2. 作業を進める過程で新しい事柄がたくさん発見される

ということです。このような場合、単に作業を進めるだけでは理想的な状態に到達できないことも多々あります。ではどうするのかというと、脇道的なコミット（`gitignore`の変更など）だけを取り出して別のプルリクエストにしたり、場合によっては仕切り直して、プルリクエストを破棄して新しいものを作り直すときもあります。これはもちろん場合によりますが、そうすることで長期的に見通しを良くすることができるかもしれません。端的に言うと、「プルリクエストと論理的な単位・意味を一致させていく」「その論理的な単位をできるだけ小さくする」ということが望ましいです。
このようなときにも、多くの場合には`git cherry-pick`のようなコマンドを使うことになります。つまり、作業を論理的に進めていくとき、先ほどの`git rebase`などと合わせて、Gitを適切に操作するスキルは必須です。

今後LLMの利用が盛んになっていくにつれて、コードの変更行数などは増えていくと予想されます。そのような場合でも、「今自分が行っていること（LLMにやらせたいこと）の意味を正しく理解する」ことは必要です。今の時点でそこに意識を向けることは、今後のために非常に有意義だと思います。そこを念頭に置いて、残りのブートキャンプでの時間を過ごしていただければと思います！

Miya096jp committed:
不要なルーティングを削除

Miya096jp commented:
@okuramasafumi 

ご助言ありがとうございます！

このPRについては、ご指摘の点について自分でも失敗だったなあと反省することが多々あり、その分得難い経験をさせて頂いたと思っています。

これ以降のIssueでは、メンバーの指摘のおかげもあり、PRやコミット単位について少しずつ意識するようになったのですが、いつも暗中模索でしたので、アドバイス頂いた考え方を今後の指針にさせていただきます！

...と気持ちを新たにしたところで残念なのですが、コンフリクトの解消がうまくいかず、本来不要であるはずのルーティングが残存してしまいました(号泣)  今更ながらですが、それだけ別のPRで修正しましたので、お手数ですがご確認をお願いいたします。なお、CIを通すために該当の修正コミットをこのPRのブランチにcherry-pickしております。

👇PRはこちらです。
#9898

komagata commented:
@Miya096jp こちら、どういう状態になっていますか？
かなり長く続いているPRなので気になっている方もいるみたいです〜。

@matuaya さんのレビューはacceptになってないのかな？

Miya096jp commented:
@komagata
お疲れ様です。

こちら@okuramasafumiさんにレビューをしていただいたので、後は最新の記事をスクレイピングし、表示を確認してから最後のPushし、@komagata
さんにmergeを依頼する予定でした。個人開発に集中しており手付かずになっていて申し訳ありません🙇🏻‍♂️  これから行います。

@okuramasafumi
routesの修正のPRも含め、レビューありがとうございました！🙇🏻‍♂️

@matuaya
お疲れ様です。
ずいぶん前のレビューの件で失礼致します。
ひょっとしたらこちらレビューして頂いたあと、まだapproveを押して頂いていないかもしれません。

お手数ですが、ご確認よろしくお願いいたします🙇🏻‍♂️

Miya096jp commented:
@komagata

お疲れ様です。
紹介・言及記事ページので公開中の全記事が、既存のcsvに全て含まれていることを確認しましたで、こちらmergeをお願いいたします。

<s>とフライングして書いてしまいましたが、rebaseして再pushしますので、少々お待ちください！</s>

@komagata
お待たせ致しました、mergeをお願いいたします。

> 公開中の全記事が、既存のcsvに全て含まれている

こちら誤りでした。1件だけ新しい記事が公開されています。その記事のURLを控えておきますので、デプロイ後の本番環境の動作確認として、管理画面からの記事登録を依頼させていただけたらと思います。よろしくお願いいたします。

Miya096jp committed:
mainのrebaseでschmema.rb/package-lock.jsonが更新

komagata commented:
@Miya096jp (CC: @okuramasafumi ) マージさせていただきました。
長期間のPRお疲れ様でした。

Miya096jp commented:
@komagata 

ありがとうございます！
とても勉強になりました🙇🏻‍♂️
